### PR TITLE
Move legacy argparser

### PR DESCRIPTION
**Pull request by Rostepher**
# Purpose

This PR is part of the staging necessary for #11827 and builds on the staging from #11880 and #11882. Put simply this PR bundles up the current argument parser and the `_apply_default_arguments` function into an `ApplyDefaultsArgumentParser` class that will call `_apply_default_arguments` anytime the `parse_args` method is called. The new parser bundles up these two interconnected behaviors making it easier to test the expected outputs.

In addition I've also taken the time to implement a new suite of unit-tests to ensure the expected behavior for the argument parser. These tests should make it easier to transition to the new system in #11827, giving reviewers more confidence that it actually works.

The unit-test suite (ab)uses Python meta classes to dynamically generate tests for each preset and argument, ensuring that they all have coverage. There's still a handful of tests types left to implement for the arguments and these tests have uncovered a preset that is broken (which will either be removed or fixed).

rdar://problem/34336890

### CODE REVIEW

**Comment by bob-wilson** [comment]
You mentioned a preset that is broken. Can you elaborate?

**Comment by bob-wilson** [comment]
@gottesmm Do you have any opinion about splitting out a build_swift package like this?

**Comment by Rostepher** [comment]
My new test suite uncovered an unsupported flag in the `buildbot_iphoneos_arm64_crosscompiler` preset, namely the `--build-swift-ios-test` option. Looks like the preset was added this year, but from my searching that option has never been available. My intuition says that the preset is not used commonly or we would have seen this crop up earlier.

**Comment by Rostepher** [comment]
As for splitting out to a new `build_swift` package I was planning to make that the defacto home for the Python sources to `build-script`. We do already have the `swift_build_support` package, but I would prefer to keep them separate and eventually phase out that package which is structured awkwardly. Any functionality necessary from `swift_build_support` will eventually be migrated over to the new `build_swift`.

I'm open to other suggestions or names for the module, I landed on `build_swift` because it was short and I think more appropriate than a module named `build_script`.

**Comment by gottesmm** [comment]
I don't have a problem with the refactoring. I would prefer the name swift_build though. We already have precedent in that area with llvm_build and swift_build_support. But let me look at this in more depth.

File: utils/build_swift/legacy_driver_arguments.py [hunk 1]
```diff
@@ -0,0 +1,1334 @@
...
+from swift_build_support.swift_build_support.targets import \
+    StdlibDeploymentTarget
+
+
+__all__ = [
+    'create_argparser',
+]
+
+
+class ApplyDefaultsArgumentParser(argparse.ArgumentParser):
```
Comment (gottesmm) [inline utils/build_swift/legacy_driver_arguments.py]: Just to loop back from our conversation, can you do this in a separate commit. I want to be able to eyeball the move of code and then see the refactoring done upon the nicely refactored code. It will make it easier for me to review.

**Review by gottesmm — CHANGES_REQUESTED** [review]

File: utils/build_swift/legacy_driver_arguments.py (hunk 1 above)
Comment (Rostepher) [reply utils/build_swift/legacy_driver_arguments.py]: Sure thing! I'll chunk out the two large code-moves into separate PRs.

Comment (gottesmm) [reply utils/build_swift/legacy_driver_arguments.py]: Thanks! = )

**Comment by Rostepher** [comment]
Alright, in order we will need to merge #11880 and then #11882 before this one can be merged. Until then the diff for this PR will look much larger since this branch has those changes in it.

**Comment by Rostepher** [comment]
@gottesmm Would you mind taking a look over this one again?

**Comment by Rostepher** [comment]
@swift-ci please smoke test

**Comment by Rostepher** [comment]
@swift-ci please smoke test

File: utils/build_swift/tests/expected_options.py [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+    """
+
+    pass
+
+
+class UnsupportedOption(_BaseOption):
+    """Option that is not supported."""
+
+    def __init__(self, *args, **kwargs):
+        kwargs['dest'] = kwargs.get('dest')
```
Comment (gottesmm) [inline utils/build_swift/tests/expected_options.py]: What are you trying to do here?

File: utils/build_swift/tests/expected_options.py [hunk 3]
```diff
@@ -0,0 +1,494 @@
...
+        super(UnsupportedOption, self).__init__(*args, **kwargs)
+
+
+# -----------------------------------------------------------------------------
+
+EXPECTED_OPTIONS = [
+    # Ignore the help options since they always call sys.exit(0)
+    _BaseOption('-h', dest='help', default=argparse.SUPPRESS),
```
Comment (gottesmm) [inline utils/build_swift/tests/expected_options.py]: Why not make a special thing for help rather than using _BaseOption.

File: utils/build_swift/tests/expected_options.py [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+    UnsupportedOption('-I'),
+
+    # FIXME: LTO flag is a special case that acts both as an option and has
+    # valid choices
+    Option('--lto', dest='lto_type'),
+    ChoicesOption('--lto', dest='lto_type', choices=['thin', 'full']),
+
+    # NOTE: We'll need to manually test the behavior of these since they
+    # validate compiler version strings.
+    _BaseOption('--clang-compiler-version',
```
Comment (gottesmm) [inline utils/build_swift/tests/expected_options.py]: Same here. Can you make a wrapper rather than using an _ class? Looks weird to me.

**Review by gottesmm — APPROVED** [review]
This looks fine to me.

File: utils/build_swift/tests/expected_options.py (hunk 2 above)
Comment (Rostepher) [reply utils/build_swift/tests/expected_options.py]: That line facilitates allowing instances of `UnsupportedOption` to skip passing in the `dest` argument since it doesn't seem necessary for that sort of option. I would be better written as `kwards['dest'] = kwargs.pop('dest', None)` which pops `dest` off the input keyword args or resolves to None and sets the value manually when passed into the superclass constructor.

File: utils/build_swift/tests/expected_options.py (hunk 3 above)
Comment (Rostepher) [reply utils/build_swift/tests/expected_options.py]: That would probably clean this up a bit. I'll make that change.

File: utils/build_swift/tests/expected_options.py (hunk 4 above)
Comment (Rostepher) [reply utils/build_swift/tests/expected_options.py]: Will do!

**Comment by Rostepher** [comment]
@swift-ci please smoke test

**Comment by Rostepher** [comment]
@swift-ci please smoke test

File: utils/build_swift/tests/expected_options.py [hunk 5]
```diff
@@ -252,16 +260,25 @@ class UnsupportedOption(_BaseOption):
     """Option that is not supported."""
 
     def __init__(self, *args, **kwargs):
-        kwargs['dest'] = kwargs.get('dest')
+        kwargs['dest'] = kwargs.pop('dest', None)
```
Comment (gottesmm) [inline utils/build_swift/tests/expected_options.py]: This is still really weird to me. What are you trying to do here?

File: utils/build_swift/tests/test_driver_arguments.py [hunk 6]
```diff
@@ -220,6 +227,8 @@ def generate_option_test(cls, option):
             return cls._generate_append_option_test(option)
         elif option.__class__ is expected_options.ChoicesOption:
             return cls._generate_choices_option_test(option)
+        elif option.__class__ is expected_options.HelpOption:
+            return cls._generate_help_option_test(option)
```
Comment (gottesmm) [inline utils/build_swift/tests/test_driver_arguments.py]: This can be done cleaner via a dictionary.

File: utils/build_swift/tests/test_driver_arguments.py [hunk 7]
```diff
@@ -143,6 +143,13 @@ def test(self):
         return test
 
     @classmethod
+    def _generate_help_option_test(cls, option):
+        def test(self):
```
Comment (gottesmm) [inline utils/build_swift/tests/test_driver_arguments.py]: This doesn't do anything, no? You aren't invoking test. What is the purpose of this?

**Review by gottesmm — COMMENTED** [review]
Some post commit review.

File: utils/build_swift/tests/test_driver_arguments.py (hunk 7 above)
Comment (Rostepher) [reply utils/build_swift/tests/test_driver_arguments.py]: The meta-class here is used to generate unit-test cases dynamically. When Python is invoked with the `unittest` argument it will execute all methods on classes subclassing `unittest.TestCase` that begin with the word `test`. It's not required in any Python unittest code to invoke the test cases yourself.

File: utils/build_swift/tests/test_driver_arguments.py (hunk 6 above)
Comment (Rostepher) [reply utils/build_swift/tests/test_driver_arguments.py]: When you say 'via a dictionary' do you mean I should have a dictionary from class to function? I'm not sure I'm following along.

File: utils/build_swift/tests/expected_options.py (hunk 5 above)
Comment (Rostepher) [reply utils/build_swift/tests/expected_options.py]: I'm injecting my own definition for `dest` such that the `UnsupportedOption` class does not need to be instantiated with it manually passed. Here I'm passing through any manually set `dest` or setting it by default to None. In all cases it's defined in the `kwargs` and satisfies the `_BaseOption` constructor.

File: utils/build_swift/tests/test_driver_arguments.py (hunk 7 above)
Comment (Rostepher) [reply utils/build_swift/tests/test_driver_arguments.py]: On a second look you're right, I forgot to return the actual test here. I'll fix this in the next PR
